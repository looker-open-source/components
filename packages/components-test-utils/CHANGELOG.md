# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.10.4](https://github.com/looker-open-source/components/compare/v0.10.3...v0.10.4) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.3](https://github.com/looker-open-source/components/compare/v0.10.2...v0.10.3) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.2](https://github.com/looker-open-source/components/compare/v0.10.1...v0.10.2) (2021-01-22)


### Bug Fixes

* On build properly set package.json "files" key ([#1806](https://github.com/looker-open-source/components/issues/1806)) ([c81c4bb](https://github.com/looker-open-source/components/commit/c81c4bb625c58ede49957aad90ac8d9d7b2c4b79))





## [0.10.1](https://github.com/looker-open-source/components/compare/v0.10.0...v0.10.1) (2021-01-22)


### Bug Fixes

* Properly build ems/cjs/ts for easy deep-imports ([#1804](https://github.com/looker-open-source/components/issues/1804)) ([2a523cd](https://github.com/looker-open-source/components/commit/2a523cd70b079944376c6dc87c18202e05a97b01))





# [0.10.0](https://github.com/looker-open-source/components/compare/v0.9.29...v0.10.0) (2021-01-19)


### Bug Fixes

* Properly configure build for tree-shaking support ([9202b72](https://github.com/looker-open-source/components/commit/9202b72bc5fb99eb3e1af7d9e56f3dc15b1df2a4))





## [0.9.15]

### Changed

- `@looker/components-test-utils` helpers will no longer produce `globalStyles`
