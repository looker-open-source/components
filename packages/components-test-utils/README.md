# Test Utilities

Shared functions used for mocking and testing Looker UI Components.
