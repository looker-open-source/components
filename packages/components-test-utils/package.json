{
  "name": "@looker/components-test-utils",
  "license": "MIT",
  "version": "0.9.12",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/looker-open-source/components",
    "directory": "packages/components-test-utils"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@looker/components-providers": "^0.9.12",
    "@testing-library/react": "^10.4.8",
    "@types/enzyme": "^3.10.5",
    "@types/jest": "^26.0.10",
    "@types/react": "^16.9.49",
    "@types/react-test-renderer": "^16.9.3",
    "enzyme": "^3.11.0",
    "jest-styled-components": "^6.3.4",
    "react": "^16.13.1",
    "react-test-renderer": "^16.13.1",
    "styled-components": "^4.4.1"
  },
  "peerDependencies": {
    "react": "^16.11"
  },
  "gitHead": "62febc162e3e45124f403e2c84fc741f68fe6714"
}
