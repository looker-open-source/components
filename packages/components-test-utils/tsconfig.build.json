{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "include": ["../../types", "src/**/*"]
}
