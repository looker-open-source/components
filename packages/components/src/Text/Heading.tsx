/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ResponsiveValue } from 'styled-system'
import {
  CompatibleHTMLProps,
  FontSizes,
  textTransform,
  TextTransformProps,
} from '@looker/design-tokens'
import { TextBase, TextBaseProps } from './TextBase'
import { TruncateProps, truncate } from './truncate'

type HeadingLevels = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

export interface HeadingProps
  extends TextBaseProps,
    TextTransformProps,
    TruncateProps,
    Omit<CompatibleHTMLProps<HTMLHeadingElement>, 'wrap'> {
  /** Heading level from h1-h6
   * @default: 'h2'
   */
  as?: HeadingLevels
}

const headingLevelFontSize = (props: HeadingProps) => {
  switch (props.as) {
    case 'h1':
      return 'xxlarge'
    case 'h3':
      return 'large'
    case 'h4':
      return 'medium'
    case 'h5':
      return 'small'
    case 'h6':
      return 'xsmall'
    case 'h2':
    default:
      return 'xlarge'
  }
}

const headingLineHeight = (props: HeadingProps): ResponsiveValue<FontSizes> =>
  props.fontSize ? props.fontSize : headingLevelFontSize(props)

export const Heading = styled(TextBase).attrs((props: HeadingProps) => ({
  fontSize: props.fontSize || headingLevelFontSize(props),
  lineHeight: props.lineHeight || headingLineHeight(props),
}))<HeadingProps>`
  ${textTransform}
  ${truncate}
`

Heading.defaultProps = {
  as: 'h2',
  fontFamily: 'brand',
  fontWeight: 'normal',
}
