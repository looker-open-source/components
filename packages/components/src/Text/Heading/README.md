# Heading

The `<Heading />` component is used to render a HTML `<h1>` - `<h6>` element, by default it will render a `<h2>` element

```tsx
function Basic() {
  return <Heading>Heading Text</Heading>;
}
```

## Heading Levels

To use a different HTML heading element, the `<Heading />` component accepts a `as` attribute that corresponds to the `<h1>` - `<h6>` elements. The font-size of each heading element maps to the [type ramp](/#!/Typography)

```tsx
function Level() {
  return <Heading as="h4">Heading Text</Heading>;
}
```

## Heading Sizes

When creating accessible pages it is important that headings create a [logical document outline](https://bitsofco.de/using-heading-elements-to-create-a-document-outline/), but sometimes the font-size of the heading element doesn't match to the needs of the design or layout. Composing the `is` and the `fontSize` attributes lets you choose the semantically correct level heading and the desired size. The available size values come from the [type ramp](/#!/Typography).

```tsx
function FontSize() {
  return <Heading fontSize="large">Heading Text</Heading>;
}
```

---

## Weight and Transform

Another common pattern for headings is to control the font-weight and the text-transform properties. The `<Heading />` component allows you to adjust those with the `fontWeight` and `textTransform` attributes.

```tsx
function FontWeight() {
  return <Heading fontWeight="bold">Heading Text</Heading>;
}
```

```tsx
function TextTransform() {
  return <Heading textTransform="uppercase">Heading Text</Heading>;
}
```

---

## Heading alignment

The `align` property allows you to adjust the `text-align` property of your `<Heading />` component. This is useful if you need to center or right align the text.

```tsx
function TextAlign() {
  return (
    <>
      <Heading textAlign="left">◀️ Align left (Default) </Heading>
      <Heading textAlign="center">◀️ Align Center ▶️</Heading>
      <Heading textAlign="right">Align Right ▶️</Heading>
    </>
  );
}
```

## Color

Heading supports a `color` prop that supports the full theme color object. Common uses are shown in the example below:

```tsx
function Color() {
  return <Heading color="text1">Heading Text</Heading>;
}
```
