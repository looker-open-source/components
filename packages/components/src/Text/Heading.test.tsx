/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { createWithTheme } from '@looker/components-test-utils'
import { Heading } from './Heading'

test('A default Heading', () => {
  const component = createWithTheme(<Heading>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading', () => {
  const component = createWithTheme(<Heading as="h1">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading sized to <h2>', () => {
  const component = createWithTheme(
    <Heading as="h1" fontSize="xlarge">
      ðŸ¥‘
    </Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading to bold', () => {
  const component = createWithTheme(
    <Heading fontSize="large" fontWeight="bold">
      ðŸ¥‘
    </Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading transformed', () => {
  const component = createWithTheme(
    <Heading textTransform="capitalize">ðŸ¥‘</Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading with variant', () => {
  const component = createWithTheme(<Heading color="subdued">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading truncated', () => {
  const component = createWithTheme(<Heading truncate>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading with multiline truncate', () => {
  const component = createWithTheme(<Heading truncateLines={2}>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
