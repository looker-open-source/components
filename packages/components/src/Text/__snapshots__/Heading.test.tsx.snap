// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <h1> Heading 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.5rem;
  font-weight: 400;
  line-height: 2.25rem;
  color: #262D33;
}

<h1
  className="c0"
  fontFamily="brand"
  fontSize="xxlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h1>
`;

exports[`A <h1> Heading sized to <h2> 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #262D33;
}

<h1
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h1>
`;

exports[`A Heading to bold 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.125rem;
  font-weight: 700;
  line-height: 1.5rem;
  color: #262D33;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="large"
  fontWeight="bold"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading transformed 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  text-transform: capitalize;
  color: #262D33;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading truncated 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #262D33;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with multiline truncate 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #262D33;
  -webkit-box-orient: vertical;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  overflow: hidden;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with variant 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #939BA5;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A default Heading 1`] = `
.c0 {
  font-family: 'Red Hat Display','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #262D33;
}

<h2
  className="c0"
  fontFamily="brand"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;
